packages/e2e-tests/test/e2e-editor.spec.ts

```diff
@@ -42,7 +44,6 @@ describe('external editor e2e', function () {
   });
 
   afterEach(async function () {
-    await TestShell.killall.call(this);
```

Comment:
Just a note, not something I have a good solution for off the top of my head, but it's now less obvious that the `afterEach` hook registered by `cleanTestShellsAfterEach()` has to run before this one – if that breaks, we'll have a flaky test here that's not going to be trivial to debug.

Reply:
> if that breaks, we'll have a flaky test here

I don't think so. If that happens the `assert` starting the test server will fail and it'll be obvious what the change needs to be 🤞

Reply:
Hm – I was thinking more of the `fs.rm()` here in this block failing because TestShell instances could still be running. I guess the existing `try/catch` accounts for that to some degree though.

Reply:
> I was thinking more of the `fs.rm()` here

Oh, I didn't understand that. Thanks for clarifying that.
We could add another check to assert no servers are running?

Reply:
I think you mean open shells, right? In any case, the added commit LGTM!

Reply:
It turns out that assert actually fails 🙈  Good catch!

I tried a fix moving the `cleanTestShellsAfterEach` call into the individual 2nd level test suites, which did solve the assert, but most DRY solution. For now, I suggest adding back the call to `killAll`.

I have ideas for future improvements though: Ideally, starting (and stopping + additional) the test shell could be declared via Mocha hooks and exposed via the Mocha Context API. This removes the need for a global "kill all" and instead just make that a logical consequence of starting a shell and it'll be easier to mix with other hooks as it'll be called on a "deeper level".

Reply:
Fwiw I agree with the idea of making test shell lifetime management more explicit (in particular, explicitly cleaning up started shells), but I'm not sure if it's a good idea to move _all_ usage of `TestShell.start()` to before hooks just from a practical point of view. A lot of our individual tests start shells with specific parameters that only really apply to that specific test, for example, and giving every single one of those tests its own mocha `context()` is probably going to impact readability and maintainability significantly.

Reply:
> A lot of our individual tests start shells with specific parameters that only really apply to that specific test, for example, and giving every single one of those tests its own mocha context() is probably going to impact readability and maintainability significantly.

I toyed around with it and reached a similar conclusion, regarding the hooks, now I'm contemplating a different API exposed on the Mocha Context which starts the shell and then registers an after hook (or event listener) to kill the shell.

Reply:
I pushed something significantly different from my original proposal and updated the description of the PR 🙈

Reply:
Should there be an `TestShell.assertNoOpenShells()` call here? :)

Reply:
You're right, I somehow mixed that up and thought it had to be here: https://github.com/mongodb-js/mongosh/blob/06018088c6de50ebd1ac3a2a18d256872cb508ae/packages/e2e-tests/test/e2e.spec.ts#L1386

I've added it into [e2e-editor.spec.ts](https://github.com/mongodb-js/mongosh/pull/2170/files#diff-8e53125a0fd24a9a2a5d85b234eb105ebd6527f4129d5691169a6606ea50c090) 👍